Force export runtime option + some tests for XmlSerializationMixin

e-kolpakov opened the pull request:
**Description:** This PR adds an option to force export any field. Forward-port of https://github.com/edx-solutions/XBlock/pull/6

The motivation is the following: `UNIQUE_ID` sentinel is used to generate a unique pseudo-random value for an XBlock field. The value generated is not actually random, and is stable (i.e. subsequent calculations return same value for given field and XBlock instances). The problem though, is that this value is not persisted in export-import workflow.

Direct motivation, more details and discussion of various approaches can be seen in the appropriate [JIRA task](https://edx-wiki.atlassian.net/browse/MCKIN-3637)

**JIRA Task:** [MCKIN-3637](https://edx-wiki.atlassian.net/browse/MCKIN-3637) and [OSPR-970](https://openedx.atlassian.net/browse/OSPR-970)
**Testing instructions:** 
No user facing parts - running unittests should be enough.

For manual verification:
1. Create an XBlock with any field that have `default=UNIQUE_ID` and `force_export=True` (or use existing one, i.e. https://github.com/open-craft/xblock-group-project-v2). 
2. Create a course, add that XBlock to course content, *do not specify a value for that field - leave it default*, but note the value somehow (i.e. display it in the student view)
3. Export course.
4. Import course. Observe the value was preserved.

edx-webhook commented:
Thanks for the pull request, @e-kolpakov! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=326&repo=edx%2FXBlock

edx-webhook commented:
Thanks for the pull request, @e-kolpakov! I've created [OSPR-970](https://openedx.atlassian.net/browse/OSPR-970) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

Kelketek commented:
:+1:

doctoryes commented on xblock/mixins.py:
```diff
@@ -22,10 +23,12 @@
+XML_NAMESPACES = OrderedDict([
+    ("option", "http://code.edx.org/xblock/option"),
+    ("block", "http://code.edx.org/xblock/block"),
+])
```
I don't see anything added in this PR that requires the `OrderedDict`. What tests require this ordering?

e-kolpakov replied:
@doctoryes [this test|https://github.com/edx/XBlock/blob/master/xblock/test/test_parsing.py#L299-L317] is flaky because namespaces *might* be put in different order.

doctoryes replied:
Makes perfect sense, thanks.

doctoryes commented:
While the addition of the `force_export` parameter seems fine (and seems needed in this situation), I'd like for the parameter and a description of its use to be added to the docstring here:

https://github.com/edx/XBlock/blob/master/xblock/fields.py#L275-L302

especially since it's directly related to the `default` parameter.

e-kolpakov commented:
@doctoryes since docstring contains constructor parameter definitions, I've converted `force_export` from runtime option to field attribute. And added it to docstring of course.

doctoryes commented:
:+1:  This PR is ready to merge. I'll do so now.

doctoryes commented:
@e-kolpakov I assume there'll be a follow-on PR for edx-platform to use this new XBlock feature?

@pdesjardins FYI - might be XBlock documentation to add here.

e-kolpakov commented:
@doctoryes so far the only use case for this is in an XBlock that runs on edx-solutions fork. So there's no need to change edx/edx-platform requirements right now. But if it is a good moment to update it I can do so. I'm in UTC+7 timezone now, so we're 12 hours apart (if you're in Boston), so I'll wait for your response on this until tomorrow morning my time. If there aren't any - I'll do a hash update PR in edx/edx-platform

doctoryes commented:
@e-kolpakov It's preferable to go ahead and expose the new functionality now to edx-platform. The alternative would be at some point in the future, someone adds newer XBlock functionality, updates edx-platform, and releases this functionality unknowingly with their own. I don't foresee problems with this XBlock change - but as a general practice, it's better to release it now.
